File: android/app/src/main/java/com/foliole/android/FolioleCompanionResourceAvailability.java
package com.foliole.android;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.FileNotFoundException;
import java.security.MessageDigest;
import java.util.HashSet;
import java.util.Set;

final class FolioleCompanionResourceAvailability {
    private FolioleCompanionResourceAvailability() {}

    static JSONObject reply(Context context, String snapshot, String body, String deviceId) throws Exception {
        JSONArray needs = new JSONObject(body).getJSONArray("resources");
        if (needs.length() > 32) throw new IllegalArgumentException("resource_availability_invalid_request");
        JSONArray claims = new JSONArray();
        Set<String> seen = new HashSet<>();
        for (int index = 0; index < needs.length(); index++) {
            JSONObject need = needs.getJSONObject(index);
            String kind = need.getString("kind"), id = need.getString("id");
            if ((!kind.equals("attachment") && !kind.equals("content_blob")) || !id.matches("[a-f0-9]{64}")
                || !seen.add(kind + ":" + id)) throw new IllegalArgumentException("resource_availability_invalid_request");
            claims.put(inspect(context, snapshot, kind, id));
        }
        return new JSONObject().put("provider_device_id", deviceId).put("resources", claims);
    }

    private static JSONObject inspect(Context context, String snapshot, String kind, String id) throws Exception {
        JSONObject claim = new JSONObject().put("kind", kind).put("id", id).put("status", "missing");
        FolioleCompanionSyncGroupResources.Resource resource;
        try {
            resource = kind.equals("attachment")
                ? FolioleCompanionSyncGroupResources.attachment(context, snapshot, id, id)
                : FolioleCompanionSyncGroupResources.contentBlob(snapshot, id);
        } catch (FileNotFoundException error) { return claim; }
        if (resource == null) return claim;
        String expectedHash = id;
        long expectedSize = resource.body.length;
        if (kind.equals("content_blob")) {
            SQLiteDatabase db = SQLiteDatabase.openDatabase(snapshot, null, SQLiteDatabase.OPEN_READONLY);
            try (Cursor row = db.rawQuery("SELECT stored_sha256, stored_size_bytes FROM content_blobs WHERE hash = ?", new String[] { id })) {
                if (!row.moveToFirst()) return claim;
                expectedHash = row.getString(0); expectedSize = row.getLong(1);
            } finally { db.close(); }
        }
        String actualHash = digest(resource.body);
        if (!actualHash.equals(expectedHash) || resource.body.length != expectedSize) return claim.put("status", "checksum_mismatch");
        return claim.put("status", "available").put("sha256", actualHash).put("size_bytes", resource.body.length);
    }

    static String digest(byte[] bytes) throws Exception {
        StringBuilder value = new StringBuilder();
        for (byte item : MessageDigest.getInstance("SHA-256").digest(bytes)) value.append(String.format("%02x", item));
        return value.toString();
    }
}

File: ios/App/App/FolioleCompanionResourceAvailability.swift
import CryptoKit
import Foundation

// sql-surface: ios-isolated-snapshot-owner
enum FolioleCompanionResourceAvailability {
    static func reply(snapshot: URL, request: Data, deviceId: String) throws -> Data {
        let payload = try JSONSerialization.jsonObject(with: request) as? [String: Any]
        guard let resources = payload?["resources"] as? [[String: Any]], resources.count <= 32 else {
            throw invalid("resource_availability_invalid_request")
        }
        var seen = Set<String>()
        var claims: [[String: Any]] = []
        for resource in resources {
            guard let kind = resource["kind"] as? String, ["attachment", "content_blob"].contains(kind),
                  let id = resource["id"] as? String,
                  id.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil,
                  seen.insert("\(kind):\(id)").inserted else { throw invalid("resource_availability_invalid_request") }
            claims.append(try inspect(snapshot: snapshot, kind: kind, id: id))
        }
        return try JSONSerialization.data(withJSONObject: ["provider_device_id": deviceId, "resources": claims])
    }

    private static func inspect(snapshot: URL, kind: String, id: String) throws -> [String: Any] {
        var result: [String: Any] = ["kind": kind, "id": id, "status": "missing"]
        let resource: FolioleCompanionSyncGroupResources.Resource?
        do {
            resource = kind == "attachment"
                ? try FolioleCompanionSyncGroupResources.attachment(snapshot: snapshot, attachmentId: id, contentHash: id)
                : try FolioleCompanionSyncGroupResources.contentBlob(snapshot: snapshot, hash: id)
        } catch let error as NSError where error.domain == NSCocoaErrorDomain && error.code == NSFileReadNoSuchFileError {
            return result
        }
        guard let resource else { return result }
        let sha256 = SHA256.hash(data: resource.body).map { String(format: "%02x", $0) }.joined()
        var expectedHash = id
        var expectedSize = resource.body.count
        if kind == "content_blob" {
            let database = try FolioleReadOnlySQLite(url: snapshot)
            guard let row = try database.rows("SELECT stored_sha256, stored_size_bytes FROM content_blobs WHERE hash = ?",
              arguments: [id]).first, let hash = row[0], let sizeText = row[1], let size = Int(sizeText) else { return result }
            expectedHash = hash
            expectedSize = size
        }
        result["status"] = sha256 == expectedHash && resource.body.count == expectedSize ? "available" : "checksum_mismatch"
        if result["status"] as? String == "available" {
            result["sha256"] = sha256
            result["size_bytes"] = resource.body.count
        }
        return result
    }

    private static func invalid(_ message: String) -> NSError {
        NSError(domain: "FolioleCompanionResourceAvailability", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
    }
}
